### PR TITLE
Refactor MEAMC potential reader

### PR DESCRIPTION
**Pull request by rbberger**
**Summary**

Refactor MEAMC potential reader code to use `PotentialFileReader`

**Related Issue(s)**

**Author(s)**

@rbberger 

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

**Implementation Notes**

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

File: src/USER-MEAMC/meam.h [hunk 1]
```diff
@@ -18,6 +18,7 @@
 
 #include <cmath>
 #include <cstring>
```
Comment (martok) [inline src/USER-MEAMC/meam.h:20]: IWYU: with str_to_lat updated, there is nothing using functions from cstring left in the header.

File: src/USER-MEAMC/pair_meamc.cpp [hunk 2]
```diff
@@ -358,272 +351,214 @@ double PairMEAMC::init_one(int i, int j)
...
+        b3[index] = values.next_double();
+        alat[index] = values.next_double();
+        esub[index] = values.next_double();
+        asub[index] = values.next_double();
+        t0[index] = values.next_double();
+        t1[index] = values.next_double();
+        t2[index] = values.next_double();
+        t3[index] = values.next_double();
+        rozero[index] = values.next_double();
+        ibar[index] = values.next_double();
```
Comment (martok) [inline src/USER-MEAMC/pair_meamc.cpp:439]: ibar is int -> values.next_int()

File: src/USER-MEAMC/pair_meamc.cpp [hunk 3]
```diff
@@ -358,272 +351,214 @@ double PairMEAMC::init_one(int i, int j)
...
+        std::string lattice_type = values.next_string();
+
+        if (!MEAM::str_to_lat(lattice_type.c_str(), true, lat[index]))
+          error->one(FLERR,fmt::format("Unrecognized lattice type in MEAM "
+                                       "library file: {}", lattice_type));
+
+        // store parameters
+
+        z[index] = values.next_double();
+        ielement[index] = values.next_double();
```
Comment (martok) [inline src/USER-MEAMC/pair_meamc.cpp:424]: ielement is int -> values.next_int()

File: src/USER-MEAMC/pair_meamc.cpp [hunk 4]
```diff
@@ -358,272 +351,214 @@ double PairMEAMC::init_one(int i, int j)
...
-
-    // params = ptrs to all fields in line
-
-    int nparams = 0;
-    params[nparams++] = strtok(line,"=(), '\t\n\r\f");
-    while (nparams < maxparams &&
-           (params[nparams++] = strtok(nullptr,"=(), '\t\n\r\f")))
-      continue;
-    nparams--;
+    ValueTokenizer values(line, "=(), '\t\n\r\f");
```
Comment (martok) [inline src/USER-MEAMC/pair_meamc.cpp:539]: Damn, I had completely forgotten how hacky this namelist "parsing" was...
(just a random thought, it's fine for this PR)

**Review by martok — CHANGES_REQUESTED** [review]
Code-based review without having compiled it since I'm away from my dev box... trusting the regtests for now.

I found two type confusions, the rest looks good to me. It's pretty amazing how much nicer modern C++ can look compared to the mostly-C-ish baseline.

**Comment by rbberger** [comment]
@martok Good catches :smile:

**Review by martok — APPROVED** [review]

**Review by akohlmey — APPROVED** [review]

**Review by sjplimp — APPROVED** [review]

**Pull request merged**